osquery/tables/system/windows/processes.cpp

```diff
@@ -147,238 +214,334 @@ Status getProcList(std::set<long>& pids) {
-      CloseHandle(tok);
-      tok = nullptr;
-    }
+  if (ret != STATUS_BUFFER_OVERFLOW && ret != STATUS_BUFFER_TOO_SMALL &&
+      ret != STATUS_INFO_LENGTH_MISMATCH) {
+    VLOG(1) << "NtQueryInformationProcess failed for " << pid << " with "
+            << ret;
+    return;
   }
 
-  if (hProcess != nullptr) {
-    CloseHandle(hProcess);
+  PVOID cmdline = static_cast<PVOID>(malloc(size_out));
+  ret = kNtQueryInformationProcess(proc, 60, cmdline, size_out, &size_out);
+  if (!NT_SUCCESS(ret)) {
+    VLOG(1) << "NtQueryInformationProcess failed for " << pid << " with "
+            << ret;
+    return;
   }
+  auto ustr = reinterpret_cast<PUNICODE_STRING>(cmdline);
+  out = wstringToString(ustr->Buffer);
 }
 
-// collect perf data into a hashmap by pid to later be refferenced
+void getProcessPathInfo(HANDLE& proc, unsigned long pid, Row& r) {
+  unsigned long out = kMaxPathSize;
+  std::shared_ptr<char> path(new char[kMaxPathSize],
+                             std::default_delete<char[]>());
+  SecureZeroMemory(path.get(), kMaxPathSize);
+  auto ret = QueryFullProcessImageName(proc, 0, path.get(), &out);
+  if (ret != TRUE) {
+    VLOG(1) << "Failed to lookup path information for process " << pid;
+    r["path"] = "-1";
+  } else {
+    r["path"] = TEXT(path.get());
+  }
 
-void genPerfPerProcess(
-    std::map<std::int32_t, std::map<std::string, std::int64_t>>& perfData) {
-  const WmiRequest request(kWinProcPerfQuery);
+  auto s = osquery::pathExists(path.get());
+  r["on_disk"] = INTEGER(s.getMessage());
 
-  if (!request.getStatus().ok()) {
-    VLOG(1) << "Failed to query process perf data from WMI";
-    return;
+  SecureZeroMemory(path.get(), kMaxPathSize);
+  if (pid == GetCurrentProcessId()) {
+    ret = GetModuleFileName(nullptr, path.get(), kMaxPathSize);
+  } else {
+    ret = GetModuleFileNameEx(proc, nullptr, path.get(), kMaxPathSize);
+  }
+  if (ret == FALSE) {
+    VLOG(1) << "Failed to get cwd for " << pid << " with " << GetLastError();
+    r["cwd"] = "-1";
+  } else {
+    r["cwd"] = TEXT(path.get());
   }
+  r["root"] = r["cwd"];
+}
 
-  const auto& results = request.results();
-  for (const auto& result : results) {
-    std::map<std::string, std::int64_t> processData;
-    long processID;
-    long handleCount = 0;
-    std::string elapsedTime;
-    std::string percentProcessorTime;
-
-    result.GetString("ElapsedTime", elapsedTime);
-    result.GetLong("HandleCount", handleCount);
-    result.GetString("PercentProcessorTime", percentProcessorTime);
-    processData["elapsed_time"] = std::stoll(elapsedTime);
-    processData["handle_count"] = handleCount;
-    processData["percent_processor_time"] = std::stoll(percentProcessorTime);
-    result.GetLong("IDProcess", processID);
-    perfData[processID] = processData;
+void genProcessUserTokenInfo(HANDLE& proc, Row& r) {
+  /// Get the process UID and GID from its SID
+  HANDLE tok = nullptr;
+  std::vector<char> tok_user(sizeof(TOKEN_USER), 0x0);
```

Comment:
Approximately how large is this? Also until now you've used smart pointers with alloced memory, is there any reason why this is different?

Reply:
Ah, so, this is a legit concern, but `TOKEN_USER` is typically very small, and considering we dynamically resize the vector based off the return of the API call, I'd actually say let's leave this as is, as it's memory safe.